### PR TITLE
Use start_test_network.sh for recovery

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves https://github.com/microsoft/CCF/issues/1095

This add a new feature to the `start_test_network.sh` script and underlying Python infra so that it is possible to recover a defunct service, i.e.:

```
$ ../start_test_network.sh -p liblogging.virtual.so
...
[2020-05-14 12:00:15.144]   Node [ 0] = 127.144.227.254:35945
[2020-05-14 12:00:15.144]   Node [ 1] = 127.29.10.59:44623
[2020-05-14 12:00:15.144]   Node [ 2] = 127.69.86.53:46645
...
^C
$ cp ./workspace/test_network_0/0.ledger .
$ cp ./workspace/test_network_0/network_enc_pubk.pem .
$ ../start_test_network.sh -p liblogging.enclave.so.signed --recover --ledger 0.ledger --network-enc-pubk network_enc_pubk.pem --common-dir ./workspace/test_network_common/ 
[2020-05-14 12:01:36.367] Starting 3 CCF nodes...
[2020-05-14 12:01:36.368] Recovering network from: 
[2020-05-14 12:01:36.368] - Ledger: 0.ledger
[2020-05-14 12:01:36.368] - Defunct network public encryption key: network_enc_pubk.pem
[2020-05-14 12:01:36.368] - Common directory: ./workspace/test_network_common/
[2020-05-14 12:01:41.148] Started CCF network with the following nodes:
[2020-05-14 12:01:41.148]   Node [ 3] = 127.48.40.201:34947
[2020-05-14 12:01:41.148]   Node [ 4] = 127.214.91.12:40041
[2020-05-14 12:01:41.148]   Node [ 5] = 127.42.131.18:34559
[2020-05-14 12:01:41.148] You can now issue business transactions to the liblogging.virtual.so application.
[2020-05-14 12:01:41.148] See https://microsoft.github.io/CCF/users/issue_commands.html for more information.
[2020-05-14 12:01:41.148] Press Ctrl+C to shutdown the network.
```

Notes:
- Most of the Python recovery capabilities have been moved to the `infra.ccf.Network` class.
- The tricky part is to reconstruct the state of the network from the point of view of the Python infra. For our existing end-to-end recovery test, we just take the state of the previous `infra.ccf.Network`. For this feature instead, we replay the ledger to find out the existing node IDs, the list of accepted/active members and the recovery threshold. The only thing we cannot find out in advance in the ID of the first recovered node (the one which will replay the ledger). Hence, the first recovery node is located at `test_network_recover_0` while the other recovered nodes are stored under their real node ID (i.e. `test_network_recover_4` and `test_network_recovery_5`).
- Added docs and relevant test in the Daily CI.

### CODE REVIEW

File: tests/infra/ccf.py [hunk 1]
```diff
@@ -275,11 +296,44 @@ def start_and_join(self, args):
         self.status = ServiceStatus.OPEN
         LOG.success("***** Network is now open *****")
 
-    def start_in_recovery(self, args, ledger_file):
-        self.common_dir = get_common_folder_name(args.workspace, args.label)
+    def start_in_recovery(self, args, ledger_file, common_dir=None):
+        self.common_dir = common_dir or get_common_folder_name(
+            args.workspace, args.label
+        )
+        self.store_current_network_encryption_key()
```
Comment (jumaffre) [inline tests/infra/ccf.py:303]: Thinking about this more, I want to make it more explicit that members have to use the network encryption public key of the previous network in order to decrypt the recovery shares. I will pass it as an extra recovery argument to `start_test_network.sh`, along with the ledger and the common directory.

File: tests/infra/member.py [hunk 2]
```diff
@@ -23,22 +23,35 @@ class MemberStatus(Enum):
...
+                f"--name={member}",
+                f"--curve={curve.name}",
+                "--gen-enc-key",
+                path=self.common_dir,
+                log_output=False,
+            ).check_returncode()
+        else:
+            # If no key generator is passed in, the identity of the member
+            # should have been created in advance (e.g. by a previous network)
+            assert os.path.isfile(
```
Comment (achamayou) [inline tests/infra/member.py:46]: fwiw, pathlib is pretty nice, but this is fine too of course.

File: tests/infra/ccf.py (hunk 1 above)
Comment (jumaffre) [reply tests/infra/ccf.py]: Done.

File: tests/infra/member.py (hunk 2 above)
Comment (jumaffre) [reply tests/infra/member.py:46]: I've stuck to `os.path` as it is what we've been using throughout the Python infra so far but we can change that in a future PR.

**Pull request merged**